x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartTrainedModelDeploymentAction.java

```diff
@@ -237,148 +236,56 @@ protected ClusterBlockException checkBlock(StartTrainedModelDeploymentAction.Req
         return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_WRITE);
     }
 
-    private static class DeploymentStartedPredicate implements Predicate<PersistentTasksCustomMetadata.PersistentTask<?>> {
+    private static class DeploymentStartedPredicate implements Predicate<TrainedModelAllocation> {
 
         private volatile Exception exception;
-        private volatile String node = "";
+
+        // for logging
+        private final String modelId;
+
+        DeploymentStartedPredicate(String modelId) {
+            this.modelId = modelId;
+        }
 
         @Override
-        public boolean test(PersistentTasksCustomMetadata.PersistentTask<?> persistentTask) {
-            if (persistentTask == null) {
-                return false;
+        public boolean test(TrainedModelAllocation trainedModelAllocation) {
+            if (trainedModelAllocation == null) {
+                // Something weird happened, it should NEVER be null...
+                return true;
             }
 
-            PersistentTasksCustomMetadata.Assignment assignment = persistentTask.getAssignment();
-
-            String reason = "__unknown__";
+            final Set<Map.Entry<String, RoutingStateAndReason>> nodesAndState = trainedModelAllocation
+                .getNodeRoutingTable()
+                .entrySet();
 
-            if (assignment != null) {
-                if (assignment.equals(JobNodeSelector.AWAITING_LAZY_ASSIGNMENT)) {
-                    return true;
+            Map<String, String> nodeFailuresAndReasons = new HashMap<>();
+            Set<String> nodesStillInitializing = new HashSet<>();
+            for (Map.Entry<String, RoutingStateAndReason> nodeIdAndState : nodesAndState) {
+                if (RoutingState.FAILED.equals(nodeIdAndState.getValue().getState())) {
+                    nodeFailuresAndReasons.put(nodeIdAndState.getKey(), nodeIdAndState.getValue().getReason());
```

Comment:
If any fail, they all fail. This can be easily changed later. We may want to support "partial allocations"

Reply:
It seems to me that it would be beneficial to our users to allow flexibility here and support partial allocations. But I'm happy to deal with this after this PR is merged. Let's raise an issue though so we don't forget.